### PR TITLE
8-bit RSSI values

### PR DESCRIPTION
**Pull request by pulquero**
Sketch that uses 8-bit RSSI values.

Preview for comments. Testing to follow.

### CODE REVIEW

File: src/node/node.ino [hunk 1]
```diff
@@ -1,846 +1,859 @@
...
+                setRxModule(settings.vtxFreq);  // Shouldn't do this in Interrupt Service Routine
+                success = true;
+                state.rxFreqSetFlag = true;
+                Serial.print(F("Set RX freq = "));
+                Serial.println(settings.vtxFreq);
+                if (settings.vtxFreq != u16val)
+                {  // if RX frequency changed
+                    writeWordToEeprom(EEPROM_ADRW_RXFREQ, settings.vtxFreq);
+                    state.nodeRssiPeak = 0;  // restart rssi peak tracking for node
+                    state.nodeRssiNadir = 999;
```
Comment (HazardCreative) [inline src/node/node.ino]:
```suggestion
                    state.nodeRssiNadir = MAX_RSSI;
```

**Comment by HazardCreative** [comment]
I like what you've done here.
I have recently made some improvements in the recalc branch that might make sense to bring in as well.

For discussion:

I'd like to understand the effect of stronger filtering. It seems to me like there's still high frequency noise in the signal that can be toned back. Switching to 8-bit RSSI would allow us to increase the sliding window size of the median filter. The current algorithm can use a maximum of 255 samples. We come out about even on memory use if we do that.

Since the system is limited to an accuracy of 1ms (only records time in millis), would we lose anything significant by throttling the RSSI polling to 1ms? This would also widen the filter's reach. The median filter doesn't introduce a delay between RSSI change and recorded time like the old weighting method did, so maybe this is viable now?

The RSSI value itself is meaningless, so why is it necessary to translate it back once we've reduced its resolution?

You make a good point about the 5808 operating at 3.3V and the Arduinos at 5V. Is there any benefit to feeding 3.3V into the Arduinos as a reference voltage? We can't make use of additional resolution, but would there be any other benefit?

**Comment by pulquero** [comment]
I don't know if it is a concern, but more samples also means more phase delay.

I translated the RSSI back so it makes sense when compared to existing enter/exit values.

I guess we could actually throttle the polling to the nyquist limit for a quad passing through a gate. That should get rid of any high frequency noise.

I checked and afaict there is no setting for 3.3v as a reference voltage (https://www.arduino.cc/reference/en/language/functions/analog-io/analogreference/).

**Comment by HazardCreative** [comment]
Some additional phased delay is fine as long as the output time is accurate. Even at 255 samples throttled to 1ms sample rate, we're only adding just over 0.1s. The delay until the exit threshold is crossed is likely to be larger, especially indoors where ~5s is not uncommon.

To get 3.3V reference, feed it in from the regulator to AREF and define analogReference as EXTERNAL.

File: src/interface/RHInterface.py [hunk 2]
```diff
@@ -79,6 +79,18 @@ def validate_checksum(data):
...
 
+def to_scaled_rssi(rssi):
+    return rssi >> 1
+
+def from_scaled_rssi(rssi):
+    return rssi << 1
+
+def unpack_rssi(node, data):
+    if node.api_level >= 18:
+        unpack_8(data)
```
Comment (HazardCreative) [inline src/interface/RHInterface.py]:
```suggestion
        return unpack_8(data)
```

File: src/interface/RHInterface.py [hunk 3]
```diff
@@ -79,6 +79,18 @@ def validate_checksum(data):
...
+    else:
+        unpack_16(data)
```
Comment (HazardCreative) [inline src/interface/RHInterface.py]:
```suggestion
        return unpack_16(data)
```

File: src/interface/RHInterface.py [hunk 4]
```diff
@@ -204,39 +219,64 @@ def update(self):
...
                                     ms_val = 0  # don't allow negative or too-large value
                                 node.lap_ms_since_start = ms_val
-                            node.node_peak_rssi = unpack_16(data[7:])
-                            node.pass_peak_rssi = unpack_16(data[9:])
-                            node.loop_time = unpack_32(data[11:])
-                            if data[15]:
+                            node.node_peak_rssi = unpack_rssi(node, data[offset_nodePeakRssi:])
+                            node.pass_peak_rssi = unpack_rssi(node, data[offset_passPeakRssi:])
+                            node.loop_time = unpack_32(data[offset_loopTime:])
+                            if data[offset_corssing]:
```
Comment (HazardCreative) [inline src/interface/RHInterface.py]:
```suggestion
                            if data[offset_crossing]:
```

File: src/interface/RHInterface.py [hunk 5]
```diff
@@ -204,39 +219,64 @@ def update(self):
                     lap_id = data[0]
                     lap_time_ms = 0
 
-                    rssi_val = unpack_16(data[5:])
+                    if node.api_level >= 18:
+                        offset_rssi = 3
```
Comment (HazardCreative) [inline src/interface/RHInterface.py]: Double check these offsets; I think they're not quite right.

File: src/node/node.ino [hunk 6]
```diff
@@ -1,846 +1,859 @@
...
+    // lap timestamp is between first and last peak RSSI
+    lastPass.timeStamp = (state.passRssiPeakRawLastTime + state.passRssiPeakRawTime) / 2;
+    lastPass.rssiNadir = state.passRssiNadir;
+    lastPass.lap = lastPass.lap + 1;
+
+    // reset lap-pass variables
+    state.crossing = false;
+    state.passRssiPeakRaw = 0;
+    state.passRssiPeak = 0;
+    state.passRssiNadir = 999;
```
Comment (HazardCreative) [inline src/node/node.ino]:
```suggestion
    state.passRssiNadir = MAX_RSSI;
```

File: src/node/node.ino [hunk 7]
```diff
@@ -1,846 +1,859 @@
...
+        case READ_ADDRESS:
+            ioBufferWrite8(i2cSlaveAddress);
+            break;
+
+        case READ_FREQUENCY:
+            ioBufferWrite16(settings.vtxFreq);
+            break;
+
+        case READ_LAP_STATS:
+        	mtime_t now = millis();
```
Comment (HazardCreative) [inline src/node/node.ino]: https://stackoverflow.com/a/23599822/2668253

File: src/node/node.ino [hunk 8]
```diff
@@ -1,846 +1,859 @@
...
+            ioBufferWrite8(lastPass.lap);
+            ioBufferWrite16(uint16_t(now - lastPass.timeStamp));  // ms since lap
+            ioBufferWriteRssi(state.rssiSmoothed);
+            ioBufferWriteRssi(state.nodeRssiPeak);
+            ioBufferWriteRssi(lastPass.rssiPeak);  // RSSI peak for last lap pass
+            ioBufferWrite32(state.loopTimeMicros);
```
Comment (HazardCreative) [inline src/node/node.ino]: Could this be reduced to a 2-byte value, too?

File: src/interface/RHInterface.py [hunk 9]
```diff
@@ -529,6 +569,18 @@ def broadcast_value_32(self, write_command, in_value):
...
+    def set_and_validate_value_rssi(self, node, write_command, read_command, in_value):
+        if node.api_level >= 18:
+            self.set_and_validate_value_8(node, write_command, read_command, to_scaled_rssi(in_value))
+        else:
+            self.set_and_validate_value_16(node, write_command, read_command, in_value)
+
+    def get_value_rssi(self, node, command):
+        if node.api_level >= 18:
+            return from_scaled_rssi(self.get_value_8(node, command))
+        else
```
Comment (HazardCreative) [inline src/interface/RHInterface.py]:
```suggestion
        else:
```

File: src/node/node.ino [hunk 10]
```diff
@@ -1,846 +1,859 @@
...
+                    writeWordToEeprom(EEPROM_ADRW_RXFREQ, settings.vtxFreq);
+                    state.nodeRssiPeak = 0;  // restart rssi peak tracking for node
+                    state.nodeRssiNadir = 999;
+                    Serial.println(F("Set nodeRssiPeak = 0, nodeRssiNadir = 999"));
+                }
+            }
+            break;
+
+        case WRITE_ENTER_AT_LEVEL:  // lap pass begins when RSSI is at or above this level
+            if (readAndValidateIoBuffer(WRITE_ENTER_AT_LEVEL, 2))
```
Comment (HazardCreative) [inline src/node/node.ino]:
```suggestion
            if (readAndValidateIoBuffer(WRITE_ENTER_AT_LEVEL, 1))
```

File: src/node/node.ino [hunk 11]
```diff
@@ -1,846 +1,859 @@
...
+            {
+                settings.enterAtLevel = ioBufferReadRssi();
+                writeWordToEeprom(EEPROM_ADRW_ENTERAT, fromScaledRssi(settings.enterAtLevel));
+                success = true;
+            }
+            break;
+
+        case WRITE_EXIT_AT_LEVEL:  // lap pass ends when RSSI goes below this level
+            if (readAndValidateIoBuffer(WRITE_EXIT_AT_LEVEL, 2))
```
Comment (HazardCreative) [inline src/node/node.ino]:
```suggestion
            if (readAndValidateIoBuffer(WRITE_EXIT_AT_LEVEL, 1))
```

File: src/node/node.ino [hunk 12]
```diff
@@ -760,55 +774,56 @@ void i2cTransmit()
...
-            ioBufferWrite32(millis() - lastPass.timeStamp);  // ms since lap
-            ioBufferWrite16(state.rssiSmoothed);
-            ioBufferWrite16(state.nodeRssiPeak);
-            ioBufferWrite16(lastPass.rssiPeak);  // RSSI peak for last lap pass
-            ioBufferWrite32(state.loopTime);
+            ioBufferWrite16(uint16_t(now - lastPass.timeStamp));  // ms since lap
+            ioBufferWriteRssi(state.rssiSmoothed);
+            ioBufferWriteRssi(state.nodeRssiPeak);
+            ioBufferWriteRssi(lastPass.rssiPeak);  // RSSI peak for last lap pass
+            ioBufferWrite32(state.loopTimeMicros);
```
Comment (pulquero) [inline src/node/node.ino]: check if can be 2-byte value

File: src/interface/RHInterface.py (hunk 5 above)
Comment (pulquero) [reply src/interface/RHInterface.py]: Wow, I was quite abit off.

**Comment by HazardCreative** [comment]
Merged as part of history-recalc.

**Pull request closed without merging**